test/integration/test_dwarf.c

```diff
@@ -86,81 +68,40 @@ bool test_dwarf3_c_basic(void) { // this should work for dwarf2 aswell
 	RzBinFile *bf = rz_bin_open(bin, "bins/elf/dwarf3_c.elf", &opt);
 	mu_assert_notnull(bf, "couldn't open file");
 
-	RzBinDwarfDebugAbbrev *da = NULL;
+	RzBinDwarfDebugAbbrevs *da = NULL;
 	// mode = 0, calls
 	// static void dump_r_bin_dwarf_debug_abbrev(FILE *f, RzBinDwarfDebugAbbrev *da)
 	// which prints out all the abbreviation
-	da = rz_bin_dwarf_parse_abbrev(bin->cur);
-	mu_assert_eq(da->count, 7, "Incorrect number of abbreviation");
-
-	// order matters
-	// I nest scopes to make it more readable, (hopefully)
-	int i = 0;
-	check_abbrev_tag(DW_TAG_compile_unit);
-	{
-		check_abbrev_children(true);
-		check_abbrev_count(8);
-		{
-			int j = 0;
-			check_abbrev_attr_name(DW_AT_producer);
-			check_abbrev_attr_form(DW_FORM_strp);
-			j++;
-			check_abbrev_attr_name(DW_AT_language);
-			check_abbrev_attr_form(DW_FORM_data1);
-			j++;
-			check_abbrev_attr_name(DW_AT_name);
-			check_abbrev_attr_form(DW_FORM_strp);
-			j++;
-			check_abbrev_attr_name(DW_AT_comp_dir);
-			check_abbrev_attr_form(DW_FORM_strp);
-			j++;
-			check_abbrev_attr_name(DW_AT_low_pc);
-			check_abbrev_attr_form(DW_FORM_addr);
-			j++;
-			check_abbrev_attr_name(DW_AT_high_pc);
-			check_abbrev_attr_form(DW_FORM_addr);
-			j++;
-			check_abbrev_attr_name(DW_AT_stmt_list);
-			check_abbrev_attr_form(DW_FORM_data4);
-		}
-	}
-	i++;
-	check_abbrev_tag(DW_TAG_variable);
-	{
-		check_abbrev_count(8);
-		check_abbrev_children(false);
-	}
-	i++;
-	check_abbrev_tag(DW_TAG_base_type);
-	{
-		check_abbrev_count(4);
-		check_abbrev_children(false);
-	}
-	i++;
-	check_abbrev_tag(DW_TAG_subprogram);
-	{
-		check_abbrev_count(12);
-		check_abbrev_children(true);
-	}
-	i++;
-	check_abbrev_tag(DW_TAG_variable);
-	{
-		check_abbrev_count(7);
-		check_abbrev_children(false);
-	}
-	i++;
-	check_abbrev_tag(DW_TAG_subprogram);
-	{
-		check_abbrev_count(10);
-		check_abbrev_children(true);
-	}
-	i++;
-	check_abbrev_tag(DW_TAG_variable);
-	{
-		check_abbrev_count(6);
-		check_abbrev_children(false);
-	}
-	i++;
+	da = rz_bin_dwarf_abbrev_parse(bin->cur);
+	mu_assert_eq(rz_bin_dwarf_abbrev_count(da), 7, "Incorrect number of abbreviation");
+
+	RzBinDwarfAbbrevTable *tbl = ht_up_find(da->tbl_by_offset, 0x0, NULL);
+	char *dump = rz_core_bin_dwarf_abbrev_decl_to_string(rz_bin_dwarf_abbrev_get(tbl, 1));
+	mu_assert_streq_free(dump, "    1      DW_TAG_compile_unit       [has children] (0x0)\n"
+				   "    DW_AT_producer                 DW_FORM_strp\n"
+				   "    DW_AT_language                 DW_FORM_data1\n"
+				   "    DW_AT_name                     DW_FORM_strp\n"
+				   "    DW_AT_comp_dir                 DW_FORM_strp\n"
+				   "    DW_AT_low_pc                   DW_FORM_addr\n"
+				   "    DW_AT_high_pc                  DW_FORM_addr\n"
+				   "    DW_AT_stmt_list                DW_FORM_data4\n",
+		"abbrev decl");
```

Comment:
Just to write tests easily. I assume they serve the same purpose?

Reply:
Not exactly. This is just a view of the data and it may change in the future. It is usually better to write unit tests by checking the various fields instead of a representation. It makes the test more robust and future proof.